File: apps/backend/app/schemas/solicitante.py
"""
Schemas Pydantic para Solicitante.
"""

from pydantic import BaseModel, EmailStr, field_validator, ConfigDict
from uuid import UUID
from datetime import datetime
import re


class SolicitanteBase(BaseModel):
    """Schema base para Solicitante."""
    
    correo: EmailStr
    nombres: str
    apellidos: str
    telefono: str | None = None
    
    @field_validator('nombres', 'apellidos')
    @classmethod
    def validate_nombres(cls, v: str) -> str:
        """Validar que nombres y apellidos solo contengan letras y espacios."""
        if not v or not v.strip():
            raise ValueError('El campo no puede estar vacío')
        
        # Permitir letras (incluyendo acentos y ñ), espacios, guiones y apóstrofes
        if not re.match(r'^[a-zA-ZáéíóúÁÉÍÓÚñÑ\s\'-]{2,100}$', v):
            raise ValueError('Solo se permiten letras, espacios, guiones y apóstrofes (2-100 caracteres)')
        
        return v.strip()
    
    @field_validator('telefono')
    @classmethod
    def validate_telefono(cls, v: str | None) -> str | None:
        """Validar formato de teléfono."""
        if v is None or v.strip() == '':
            return None
        
        # Remover espacios y guiones para validación
        telefono_clean = v.replace(' ', '').replace('-', '')
        
        # Validar que solo contenga dígitos y tenga entre 7 y 20 caracteres
        if not re.match(r'^\d{7,20}$', telefono_clean):
            raise ValueError('El teléfono debe contener entre 7 y 20 dígitos')
        
        return telefono_clean


class SolicitanteCreate(SolicitanteBase):
    """Schema para crear un Solicitante."""
    pass


class SolicitanteUpdate(BaseModel):
    """Schema para actualizar un Solicitante."""
    
    correo: EmailStr | None = None
    nombres: str | None = None
    apellidos: str | None = None
    telefono: str | None = None
    
    @field_validator('nombres', 'apellidos')
    @classmethod
    def validate_nombres(cls, v: str | None) -> str | None:
        """Validar que nombres y apellidos solo contengan letras y espacios."""
        if v is None:
            return None
        
        if not v.strip():
            raise ValueError('El campo no puede estar vacío')
        
        if not re.match(r'^[a-zA-ZáéíóúÁÉÍÓÚñÑ\s\'-]{2,100}$', v):
            raise ValueError('Solo se permiten letras, espacios, guiones y apóstrofes (2-100 caracteres)')
        
        return v.strip()
    
    @field_validator('telefono')
    @classmethod
    def validate_telefono(cls, v: str | None) -> str | None:
        """Validar formato de teléfono."""
        if v is None or v.strip() == '':
            return None
        
        telefono_clean = v.replace(' ', '').replace('-', '')
        
        if not re.match(r'^\d{7,20}$', telefono_clean):
            raise ValueError('El teléfono debe contener entre 7 y 20 dígitos')
        
        return telefono_clean


class SolicitanteResponse(SolicitanteBase):
    """Schema para respuesta de Solicitante."""
    
    id: UUID
    created_at: datetime
    updated_at: datetime
    
    model_config = ConfigDict(from_attributes=True)

File: apps/frontend/portal-externo/src/schemas/solicitante.ts
import { z } from 'zod';

/**
 * Schema de validación para Solicitante
 */
export const solicitanteSchema = z.object({
  correo: z
    .string()
    .min(5, 'El correo debe tener al menos 5 caracteres')
    .email('Formato de correo inválido')
    .toLowerCase()
    .trim(),
  
  nombres: z
    .string()
    .min(2, 'Los nombres deben tener al menos 2 caracteres')
    .max(100, 'Los nombres no pueden exceder 100 caracteres')
    .regex(/^[a-zA-ZáéíóúÁÉÍÓÚñÑ\s]+$/, 'Solo se permiten letras y espacios')
    .trim()
    .transform(val => val.replace(/\s+/g, ' ')), // Normalizar espacios múltiples
  
  apellidos: z
    .string()
    .min(2, 'Los apellidos deben tener al menos 2 caracteres')
    .max(100, 'Los apellidos no pueden exceder 100 caracteres')
    .regex(/^[a-zA-ZáéíóúÁÉÍÓÚñÑ\s]+$/, 'Solo se permiten letras y espacios')
    .trim()
    .transform(val => val.replace(/\s+/g, ' ')),
  
  telefono: z
    .string()
    .regex(/^\d{7,20}$/, 'El teléfono debe tener entre 7 y 20 dígitos')
    .optional()
    .or(z.literal('')),
});

/**
 * Tipo inferido del schema
 */
export type SolicitanteFormData = z.infer<typeof solicitanteSchema>;
